File: lambdas/algorithms/searching/java-searching/src/main/java/InterpolationSearch.java
public class InterpolationSearch extends BaseSearch {

    @Override
    protected int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;

        while (low <= high && target >= nums[low] && target <= nums[high]) {
            iterations++;
            comparisons += 3; // target >= nums[low] && target <= nums[high]

            if (low == high) {
                comparisons++; // nums[low] == target
                if (nums[low] == target) return low;
                return -1;
            }

            int pos = low + (((high - low) *
                    (target - nums[low])) / (nums[high] - nums[low]));

            comparisons++; // nums[pos] == target
            if (nums[pos] == target) return pos;

            comparisons++; // nums[pos] < target
            if (nums[pos] < target) low = pos + 1;
            else high = pos - 1;
        }

        return -1;
    }

}

File: lambdas/algorithms/searching/python-searching/src/algorithms/interpolation_search.py
from src.base_search import BaseSearch
from typing import List

class InterpolationSearch(BaseSearch):
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        
        while left <= right and target >= nums[left] and target <= nums[right]:
            self.iterations += 1
            
            # interpolation search
            if right == left:
                pos = left
            else:
                pos = left + ((right - left) * (target - nums[left])) // (nums[right] - nums[left])
            
            if pos < 0 or pos >= len(nums):
                break
                
            self.comparisons += 1
            if nums[pos] == target:
                return pos
            
            self.comparisons += 1
            if nums[pos] < target:
                left = pos + 1
            else:
                right = pos - 1
                
        return -1